File: tests/game_objects/test_attributes.py
"""Tests all functions in game_objects/attributes.py."""
from __future__ import annotations

# Builtin
from typing import cast

# Pip
import pytest

# Custom
from hades.game_objects.attributes import (
    GameObjectAttributeBase,
    GameObjectAttributeError,
    deal_damage,
)
from hades.game_objects.base import ComponentType
from hades.game_objects.system import ECS

__all__ = ()


class FullGameObjectAttribute(GameObjectAttributeBase):
    """Represents a full game object attribute useful for testing."""

    # Class variables
    component_type: ComponentType = ComponentType.HEALTH


class EmptyGameObjectAttribute(GameObjectAttributeBase):
    """Represents an empty game object attribute useful for testing."""

    # Class variables
    component_type: ComponentType = ComponentType.ARMOUR
    instant_effect: bool = False
    maximum: bool = False
    status_effect: bool = False
    upgradable: bool = False


@pytest.fixture()
def ecs() -> ECS:
    """Create an entity component system for use in testing.

    Returns:
        The entity component system for use in testing.
    """
    return ECS()


@pytest.fixture()
def full_game_object_attribute(ecs: ECS) -> FullGameObjectAttribute:
    """Create a full game object attribute for use in testing.

    Args:
        ecs: The entity component system for use in testing.

    Returns:
        The full game object attribute for use in testing.
    """
    ecs.add_game_object(
        {"attributes": {ComponentType.HEALTH: (150, 3)}},
        FullGameObjectAttribute,
    )
    return cast(
        FullGameObjectAttribute,
        ecs.get_component_for_game_object(0, ComponentType.HEALTH),
    )


@pytest.fixture()
def empty_game_object_attribute(ecs: ECS) -> EmptyGameObjectAttribute:
    """Create an empty game object attribute for use in testing.

    Args:
        ecs: The entity component system for use in testing.

    Returns:
        The empty game object attribute for use in testing.
    """
    ecs.add_game_object(
        {"attributes": {ComponentType.ARMOUR: (100, 5)}},
        EmptyGameObjectAttribute,
    )
    return cast(
        EmptyGameObjectAttribute,
        ecs.get_component_for_game_object(0, ComponentType.ARMOUR),
    )


@pytest.fixture()
def both_game_object_attributes(
    ecs: ECS,
) -> tuple[FullGameObjectAttribute, EmptyGameObjectAttribute]:
    """Create two game object attributes for use in testing.

    Args:
        ecs: The entity component system for use in testing.

    Returns:
        Both game object attributes for use in testing.
    """
    ecs.add_game_object(
        {
            "attributes": {
                ComponentType.HEALTH: (200, 4),
                ComponentType.ARMOUR: (150, 6),
            },
        },
        FullGameObjectAttribute,
        EmptyGameObjectAttribute,
    )
    return cast(
        FullGameObjectAttribute,
        ecs.get_component_for_game_object(0, ComponentType.HEALTH),
    ), cast(
        EmptyGameObjectAttribute,
        ecs.get_component_for_game_object(0, ComponentType.ARMOUR),
    )


def test_raise_game_object_attribute_error() -> None:
    """Test that GameObjectAttributeError is raised correctly."""
    with pytest.raises(
        expected_exception=GameObjectAttributeError,
        match="The game object attribute `test` cannot be upgraded.",
    ):
        raise GameObjectAttributeError(name="test", error="be upgraded")


def test_full_game_object_attribute_init(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a full game object attribute is initialised correctly.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    assert (
        repr(full_game_object_attribute)
        == "<FullGameObjectAttribute (Value=150) (Max value=150) (Level=0/3)>"
    )


def test_empty_game_object_attribute_init(
    empty_game_object_attribute: EmptyGameObjectAttribute,
) -> None:
    """Test that an empty game object attribute is initialised correctly.

    Args:
        empty_game_object_attribute: The empty game object attribute for use in testing.
    """
    assert (
        repr(empty_game_object_attribute)
        == "<EmptyGameObjectAttribute (Value=100) (Max value=inf) (Level=0/5)>"
    )


def test_full_game_object_attribute_setter_lower(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a full game object attribute is set with a lower value correctly.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.value = 100
    assert full_game_object_attribute.value == 100


def test_full_game_object_attribute_setter_higher(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a full game object attribute is set with a higher value correctly.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.value = 200
    assert full_game_object_attribute.value == 150


def test_full_game_object_attribute_setter_isub(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that subtracting a value from the full game object attribute is correct.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.value -= 200
    assert full_game_object_attribute.value == 0


def test_full_game_object_attribute_setter_iadd(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that adding a value to the full game object attribute is correct.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.value += 100
    assert full_game_object_attribute.value == 150


def test_full_game_object_attribute_upgrade_value_equal_max(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a full game object attribute is upgraded correctly.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    assert full_game_object_attribute.upgrade(lambda level: 150 * (level + 1))
    assert full_game_object_attribute.value == 300
    assert full_game_object_attribute.max_value == 300
    assert full_game_object_attribute.current_level == 1


def test_full_game_object_attribute_upgrade_value_lower_max(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a full game object attribute is upgraded if value is lower than max.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.value -= 50
    assert full_game_object_attribute.upgrade(lambda level: 150 + 2 ^ level)
    assert full_game_object_attribute.value == 101
    assert full_game_object_attribute.max_value == 151
    assert full_game_object_attribute.current_level == 1


def test_full_game_object_attribute_upgrade_max_limit(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a full game object attribute is not upgraded if level limit is reached.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.upgrade(lambda _: 0)
    full_game_object_attribute.upgrade(lambda _: 0)
    full_game_object_attribute.upgrade(lambda _: 0)
    assert not full_game_object_attribute.upgrade(lambda _: 0)


def test_full_game_object_attribute_upgrade_invalid_increase(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a full game object attribute is not upgraded given an invalid increase.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    with pytest.raises(expected_exception=TypeError):
        full_game_object_attribute.upgrade(
            lambda _: "str",  # type: ignore[arg-type,return-value]
        )


def test_full_game_object_attribute_apply_instant_effect_lower(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that an instant effect is applied if the value is lower than the max.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.value -= 50
    assert full_game_object_attribute.apply_instant_effect(lambda level: 10 * level, 2)
    assert full_game_object_attribute.value == 120


def test_full_game_object_attribute_apply_instant_effect_equal(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that an instant effect is not applied if the value is equal to the max.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    assert not full_game_object_attribute.apply_instant_effect(lambda _: 50, 3)
    assert full_game_object_attribute.value == 150


def test_full_game_object_attribute_apply_status_effect_no_applied_effect(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a status effect is applied if no status effect is currently applied.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    assert full_game_object_attribute.apply_status_effect(
        lambda level: 150 + 3**level,
        lambda level: 20 + 10 * level,
        2,
    )
    assert (
        repr(full_game_object_attribute.applied_status_effect)
        == "StatusEffect(value=159, duration=40, original_value=150,"
        " original_max_value=150, time_counter=0)"
    )


def test_full_game_object_attribute_apply_status_effect_value_lower_max(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a status effect is applied if the value is lower than the max.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.value -= 20
    assert full_game_object_attribute.apply_status_effect(
        lambda level: 20 * level,
        lambda level: 10 - 2**level,
        3,
    )
    assert (
        repr(full_game_object_attribute.applied_status_effect)
        == "StatusEffect(value=60, duration=2, original_value=130,"
        " original_max_value=150, time_counter=0)"
    )


def test_full_game_object_attribute_apply_status_effect_existing_status_effect(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a status effect is not applied if a status effect is already applied.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.apply_status_effect(lambda _: 50, lambda _: 20, 3)
    assert not full_game_object_attribute.apply_status_effect(
        lambda _: 60,
        lambda _: 30,
        2,
    )


def test_full_game_object_attribute_on_update_no_deltatime(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a game object is updated when no time has passed.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.apply_status_effect(
        lambda level: level * 2,
        lambda level: level + 100,
        2,
    )
    full_game_object_attribute.on_update(0)
    assert full_game_object_attribute.applied_status_effect
    assert full_game_object_attribute.applied_status_effect.time_counter == 0


def test_full_game_object_attribute_on_update_larger_deltatime(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a status effect is removed if deltatime is larger than the duration.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.apply_status_effect(
        lambda level: 2**level,
        lambda _: 20,
        2,
    )
    full_game_object_attribute.on_update(30)
    assert full_game_object_attribute.value == 150
    assert full_game_object_attribute.max_value == 150
    assert not full_game_object_attribute.applied_status_effect


def test_full_game_object_attribute_on_update_multiple_deltatimes(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a status effect is removed after multiple updates.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    # Apply the status effect and check the attribute's state after one update
    full_game_object_attribute.apply_status_effect(
        lambda level: 50 + level / 2,
        lambda level: 3**level + 50,
        3,
    )
    full_game_object_attribute.on_update(40)
    assert full_game_object_attribute.applied_status_effect
    assert full_game_object_attribute.applied_status_effect.time_counter == 40
    assert full_game_object_attribute.value == 201.5
    assert full_game_object_attribute.max_value == 201.5

    # Check the attribute's state after the final update
    full_game_object_attribute.on_update(40)
    assert full_game_object_attribute.value == 150
    assert full_game_object_attribute.max_value == 150
    assert not full_game_object_attribute.applied_status_effect


def test_full_game_object_attribute_on_update_less_than_original(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a status effect is removed when value is less than the original.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    full_game_object_attribute.apply_status_effect(lambda _: 100, lambda _: 20, 4)
    full_game_object_attribute.value -= 200
    full_game_object_attribute.on_update(30)
    assert full_game_object_attribute.value == 50
    assert full_game_object_attribute.max_value == 150
    assert not full_game_object_attribute.applied_status_effect


def test_full_game_object_attribute_on_update_no_status_effect(
    full_game_object_attribute: FullGameObjectAttribute,
) -> None:
    """Test that a game object is updated if a status effect does not exist.

    Args:
        full_game_object_attribute: The full game object attribute for use in testing.
    """
    assert not full_game_object_attribute.applied_status_effect
    full_game_object_attribute.on_update(5)
    assert not full_game_object_attribute.applied_status_effect


def test_empty_game_object_attribute_setter(
    empty_game_object_attribute: EmptyGameObjectAttribute,
) -> None:
    """Test that an empty game object attribute's value can be changed.

    Args:
        empty_game_object_attribute: The empty game object attribute for use in testing.
    """
    empty_game_object_attribute.value -= 10
    assert empty_game_object_attribute.value == 90


def test_empty_game_object_attribute_upgrade(
    empty_game_object_attribute: EmptyGameObjectAttribute,
) -> None:
    """Test that an empty game object attribute raises an error when upgraded.

    Args:
        empty_game_object_attribute: The empty game object attribute for use in testing.
    """
    with pytest.raises(
        expected_exception=GameObjectAttributeError,
        match=(
            "The game object attribute `EmptyGameObjectAttribute` cannot be upgraded."
        ),
    ):
        empty_game_object_attribute.upgrade(lambda _: 0)


def test_empty_game_object_attribute_instant_effect(
    empty_game_object_attribute: EmptyGameObjectAttribute,
) -> None:
    """Test that an instant effect raises an error on an empty game object attribute.

    Args:
        empty_game_object_attribute: The empty game object attribute for use in testing.
    """
    with pytest.raises(
        expected_exception=GameObjectAttributeError,
        match=(
            "The game object attribute `EmptyGameObjectAttribute` cannot have an"
            " instant effect."
        ),
    ):
        empty_game_object_attribute.apply_instant_effect(lambda _: 0, 5)


def test_empty_game_object_attribute_apply_status_effect(
    empty_game_object_attribute: EmptyGameObjectAttribute,
) -> None:
    """Test that a status effect raises an error on an empty game object attribute.

    Args:
        empty_game_object_attribute: The empty game object attribute for use in testing.
    """
    with pytest.raises(
        expected_exception=GameObjectAttributeError,
        match=(
            "The game object attribute `EmptyGameObjectAttribute` cannot have a status"
            " effect."
        ),
    ):
        empty_game_object_attribute.apply_status_effect(lambda _: 0, lambda _: 0, 6)


def test_empty_game_object_attribute_on_update(
    empty_game_object_attribute: EmptyGameObjectAttribute,
) -> None:
    """Test that updating a game object fails on an empty game object attribute.

    Args:
        empty_game_object_attribute: The empty game object attribute for use in testing.
    """
    assert not empty_game_object_attribute.applied_status_effect
    empty_game_object_attribute.on_update(1)
    assert not empty_game_object_attribute.applied_status_effect


def test_deal_damage_low_health_armour(
    ecs: ECS,
    both_game_object_attributes: tuple[
        FullGameObjectAttribute,
        EmptyGameObjectAttribute,
    ],
) -> None:
    """Test that damage is dealt when health and armour are lower than damage.

    Args:
        ecs: The entity component system for use in testing.
        both_game_object_attributes: Both game object attributes for use in testing.
    """
    both_game_object_attributes[0].value = 150
    both_game_object_attributes[1].value = 100
    deal_damage(0, ecs, 200)
    assert both_game_object_attributes[0].value == 50
    assert both_game_object_attributes[1].value == 0


def test_deal_damage_large_armour(
    ecs: ECS,
    both_game_object_attributes: tuple[
        FullGameObjectAttribute,
        EmptyGameObjectAttribute,
    ],
) -> None:
    """Test that no damage is dealt when armour is larger than damage.

    Args:
        ecs: The entity component system for use in testing.
        both_game_object_attributes: Both game object attributes for use in testing.
    """
    deal_damage(0, ecs, 100)
    assert both_game_object_attributes[0].value == 200
    assert both_game_object_attributes[1].value == 50


def test_deal_damage_zero_damage(
    ecs: ECS,
    both_game_object_attributes: tuple[
        FullGameObjectAttribute,
        EmptyGameObjectAttribute,
    ],
) -> None:
    """Test that no damage is dealt when damage is zero.

    Args:
        ecs: The entity component system for use in testing.
        both_game_object_attributes: Both game object attributes for use in testing.
    """
    deal_damage(0, ecs, 0)
    assert both_game_object_attributes[0].value == 200
    assert both_game_object_attributes[1].value == 150


def test_deal_damage_zero_armour(
    ecs: ECS,
    both_game_object_attributes: tuple[
        FullGameObjectAttribute,
        EmptyGameObjectAttribute,
    ],
) -> None:
    """Test that damage is dealt when armour is zero.

    Args:
        ecs: The entity component system for use in testing.
        both_game_object_attributes: Both game object attributes for use in testing.
    """
    both_game_object_attributes[1].value = 0
    deal_damage(0, ecs, 100)
    assert both_game_object_attributes[0].value == 100
    assert both_game_object_attributes[1].value == 0


def test_deal_damage_zero_health(
    ecs: ECS,
    both_game_object_attributes: tuple[
        FullGameObjectAttribute,
        EmptyGameObjectAttribute,
    ],
) -> None:
    """Test that damage is dealt when health is zero.

    Args:
        ecs: The entity component system for use in testing.
        both_game_object_attributes: Both game object attributes for use in testing.
    """
    both_game_object_attributes[0].value = 0
    deal_damage(0, ecs, 200)
    assert both_game_object_attributes[0].value == 0
    assert both_game_object_attributes[1].value == 0


def test_deal_damage_nonexistent_attributes(ecs: ECS) -> None:
    """Test that no damage is dealt when the attributes are not initialised.

    Args:
        ecs: The entity component system for use in testing.
    """
    ecs.add_game_object({})
    with pytest.raises(expected_exception=KeyError):
        deal_damage(0, ecs, 100)

File: hades_extensions/test/game_objects/systems/test_attributes.cpp
// External includes
#include "gtest/gtest.h"

// Custom includes
#include "macros.hpp"
#include "game_objects/systems/attributes.hpp"

// ----- CLASSES ------------------------------
/// Represents a full game object attribute useful for testing.
class FullGameObjectAttribute : public GameObjectAttributeBase {
 public:
  /// Initialise the object.
  ///
  /// @param initial_value - The initial value of the view distance attribute.
  /// @param level_limit - The level limit of the view distance attribute.
  FullGameObjectAttribute(double initial_value, int level_limit) : GameObjectAttributeBase(initial_value,
                                                                                           level_limit) {}
};

/// Represents an empty game object attribute useful for testing.
class EmptyGameObjectAttribute : public GameObjectAttributeBase {
 public:
  /// Initialise the object.
  ///
  /// @param initial_value - The initial value of the view distance attribute.
  /// @param level_limit - The level limit of the view distance attribute.
  EmptyGameObjectAttribute(double initial_value, int level_limit) : GameObjectAttributeBase(initial_value,
                                                                                            level_limit) {}

  /// Get if the game object attribute can have instant effects or not.
  ///
  /// @return Whether the game object attribute can have instant effects or not.
  [[nodiscard]] bool has_instant_effect() const final { return false; }

  /// Get if the game object attribute has a maximum value or not.
  ///
  /// @return Whether the game object attribute has a maximum value or not.
  [[nodiscard]] bool has_maximum() const final { return false; }

  /// Get if the game object attribute can have status effects or not.
  ///
  /// @return Whether the game object attribute can have status effects or not.
  [[nodiscard]] bool has_status_effect() const final { return false; }

  /// Get if the game object attribute can be upgraded or not.
  ///
  /// @return Whether the game object attribute can be upgraded or not.
  [[nodiscard]] bool is_upgradable() const final { return false; }
};

// ----- FIXTURES ------------------------------
/// Implements the GameObjectAttribute fixture for the game_objects/systems/attacks.hpp tests.
class GameObjectAttributeFixture : public testing::Test {
 protected:
  /// The registry that manages the game objects, components, and systems.
  Registry registry{};

  /// Set up the fixture for the tests.
  void SetUp() override {
    registry.add_system(std::make_shared<GameObjectAttributeSystem>(registry));
  }

  /// Create two game object attributes for use in testing.
  void create_game_object_attributes() {
    std::vector<std::unique_ptr<ComponentBase>> components;
    components.push_back(std::make_unique<FullGameObjectAttribute>(150, 3));
    components.push_back(std::make_unique<EmptyGameObjectAttribute>(100, 5));
    registry.create_game_object(false, std::move(components));
  }

  /// Get the game object attribute system from the registry.
  ///
  /// @return The game object attribute system.
  std::shared_ptr<GameObjectAttributeSystem> get_game_object_attribute_system() {
    return registry.find_system<GameObjectAttributeSystem>();
  }
};

// ----- TESTS ----------------------------------
/// Test that a game object is updated correctly with a small delta time.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemUpdateSmallDeltaTime) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>>
      status_effect = std::make_tuple([](int level) { return level * 2; }, [](int level) { return level + 100; });
  get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 2);
  get_game_object_attribute_system()->update(5);
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  ASSERT_EQ(full_game_object_attribute->value(), 154);
  ASSERT_EQ(full_game_object_attribute->max_value(), 154);
  ASSERT_EQ(full_game_object_attribute->applied_status_effect->time_counter, 5);
}

/// Test that a status effect is removed with a large delta time.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemUpdateLargeDeltaTime) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>>
      status_effect = std::make_tuple([](int level) { return std::pow(level, 2); }, [](int level) { return 20; });
  get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 2);
  get_game_object_attribute_system()->update(30);
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  ASSERT_EQ(full_game_object_attribute->value(), 150);
  ASSERT_EQ(full_game_object_attribute->max_value(), 150);
  ASSERT_FALSE(full_game_object_attribute->applied_status_effect.has_value());
}

/// Test that a status effect is removed after multiple updates.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemUpdateMultipleDeltaTimes) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>> status_effect =
      std::make_tuple([](int level) { return 50 + level / 2; }, [](int level) { return std::pow(level, 3) + 50; });
  get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 3);
  get_game_object_attribute_system()->update(40);
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  ASSERT_EQ(full_game_object_attribute->value(), 201.5);
  ASSERT_EQ(full_game_object_attribute->max_value(), 201.5);
  ASSERT_EQ(full_game_object_attribute->applied_status_effect->time_counter, 40);
  get_game_object_attribute_system()->update(40);
  ASSERT_EQ(full_game_object_attribute->value(), 150);
  ASSERT_EQ(full_game_object_attribute->max_value(), 150);
  ASSERT_FALSE(full_game_object_attribute->applied_status_effect.has_value());
}

/// Test that a status effect is removed when value is less than the original.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemUpdateLessThanOriginal) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>>
      status_effect = std::make_tuple([](int level) { return 100; }, [](int level) { return 20; });
  get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 4);
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  full_game_object_attribute->value(full_game_object_attribute->value() - 200);
  get_game_object_attribute_system()->update(30);
  ASSERT_EQ(full_game_object_attribute->value(), 50);
  ASSERT_EQ(full_game_object_attribute->max_value(), 150);
  ASSERT_FALSE(full_game_object_attribute->applied_status_effect.has_value());
}

/// Test that a game object is updated if a status effect does not exist.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemUpdateNoStatusEffect) {
  create_game_object_attributes();
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  ASSERT_FALSE(full_game_object_attribute->applied_status_effect.has_value());
  get_game_object_attribute_system()->update(5);
  ASSERT_FALSE(full_game_object_attribute->applied_status_effect.has_value());
}

/// Test that updating a game object fails on an empty game object attribute.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemUpdateEmptyGameObjectAttribute) {
  create_game_object_attributes();
  auto empty_game_object_attribute = registry.get_component<EmptyGameObjectAttribute>(0);
  ASSERT_FALSE(empty_game_object_attribute->applied_status_effect.has_value());
  get_game_object_attribute_system()->update(5);
  ASSERT_FALSE(empty_game_object_attribute->applied_status_effect.has_value());
}

// TODO: Reformat this (need to get clang-format ready for this)

/// Test that an instant effect is not applied if the value is equal to the max.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyInstantEffectEqual) {
  create_game_object_attributes();
  ASSERT_FALSE(get_game_object_attribute_system()->apply_instant_effect<FullGameObjectAttribute>(0,
                                                                                                 [](int level) { return 50; },
                                                                                                 3));
  ASSERT_EQ(registry.get_component<FullGameObjectAttribute>(0)->value(), 150);
}

/// Test that an instant effect is applied if the value is lower than the max.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyInstantEffectLower) {
  create_game_object_attributes();
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  full_game_object_attribute->value(full_game_object_attribute->value() - 50);
  ASSERT_TRUE(get_game_object_attribute_system()->apply_instant_effect<FullGameObjectAttribute>(0, [](int level) {
                                                                                                  return 10 * level;
                                                                                                },
                                                                                                2));
  ASSERT_EQ(full_game_object_attribute->value(), 120);
}

/// Test that an empty game object attribute fails when applied an instant effect.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyInstantEffectEmptyGameObjectAttribute) {
  create_game_object_attributes();
  ASSERT_FALSE(get_game_object_attribute_system()->apply_instant_effect<EmptyGameObjectAttribute>(0,
                                                                                                  [](int level) { return 0; },
                                                                                                  3));
}

/// Test that an exception is raised if an invalid game object ID is provided.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyInstantEffectInvalidGameObjectId) {
  create_game_object_attributes();
  ASSERT_THROW_MESSAGE(get_game_object_attribute_system()->apply_instant_effect<FullGameObjectAttribute>(-1,
                                                                                                         [](int level) { return 0; },
                                                                                                         3),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
}

/// Test that a status effect is applied if no status effect is currently applied.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyStatusEffectNoAppliedEffect) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>> status_effect =
      std::make_tuple([](int level) { return 150 + std::pow(level, 3); }, [](int level) { return 20 + 10 * level; });
  ASSERT_TRUE(get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 2));
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  ASSERT_EQ(full_game_object_attribute->value(), 308);
  ASSERT_EQ(full_game_object_attribute->max_value(), 308);
  auto applied_status_effect = full_game_object_attribute->applied_status_effect.value();
  ASSERT_EQ(applied_status_effect.value, 158);
  ASSERT_EQ(applied_status_effect.duration, 40);
  ASSERT_EQ(applied_status_effect.original_value, 150);
  ASSERT_EQ(applied_status_effect.original_max_value, 150);
  ASSERT_EQ(applied_status_effect.time_counter, 0);
}

/// Test that a status effect is applied if the value is lower than the max.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyStatusEffectValueLowerMax) {
  create_game_object_attributes();
  auto full_game_object_attribute = registry.get_component<FullGameObjectAttribute>(0);
  full_game_object_attribute->value(full_game_object_attribute->value() - 20);
  std::tuple<std::function<double(int)>, std::function<double(int)>> status_effect =
      std::make_tuple([](int level) { return 20 * level; }, [](int level) { return 10 - std::pow(level, 2); });
  ASSERT_TRUE(get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 3));
  ASSERT_TRUE(full_game_object_attribute->applied_status_effect.has_value());
}

/// Test that a status effect is not applied if a status effect is already applied.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyStatusEffectExistingStatusEffect) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>>
      status_effect = std::make_tuple([](int level) { return 0; }, [](int level) { return 0; });
  get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 3);
  ASSERT_TRUE(registry.get_component<FullGameObjectAttribute>(0)->applied_status_effect.has_value());
  ASSERT_FALSE(get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(0, status_effect, 2));
}

/// Test that an empty game object attribute fails when applied a status effect.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyStatusEffectEmptyGameObjectAttribute) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>>
      status_effect = std::make_tuple([](int level) { return 0; }, [](int level) { return 0; });
  ASSERT_FALSE(get_game_object_attribute_system()->apply_status_effect<EmptyGameObjectAttribute>(0, status_effect, 3));
}

/// Test that an exception is raised if an invalid game object ID is provided.
TEST_F(GameObjectAttributeFixture, TestGameObjectAttributeSystemApplyStatusEffectInvalidGameObjectId) {
  create_game_object_attributes();
  std::tuple<std::function<double(int)>, std::function<double(int)>>
      status_effect = std::make_tuple([](int level) { return 0; }, [](int level) { return 0; });
  ASSERT_THROW_MESSAGE(get_game_object_attribute_system()->apply_status_effect<FullGameObjectAttribute>(-1,
                                                                                                        status_effect,
                                                                                                        3),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
}
